File: temps/modules/alenio-thermoworks/src/AlenioThermoworksModule.ts
import {
  NativeModule,
  requireNativeModule,
  type EventSubscription,
} from "expo-modules-core";

export type ThermoworksInitResult = {
  ok: boolean;
  sdkVersion: string;
  bluetoothAvailable: boolean;
  error?: string;
};

export type ThermoworksDiagnostics = {
  module: "alenio-thermoworks";
  platform: "ios" | "android" | "unknown";
  sdkVersion: string | null;
  available: boolean;
  initialized: boolean;
  bluetoothAvailable: boolean;
  scanning?: boolean;
  discoveredCount?: number;
  connectedDeviceId?: string | null;
  error?: string;
};

export type ThermoworksDeviceType =
  | "PEN_BLUE"
  | "TEMPTEST_BLUE"
  | "UNKNOWN"
  | "OTHER";

export type ThermoworksDiscoveredDevice = {
  deviceId: string;
  name: string;
  rssi?: number;
  serialNumber?: string;
  deviceType: ThermoworksDeviceType;
};

export type ThermoworksDevicesEvent = {
  type: "devices";
  devices: ThermoworksDiscoveredDevice[];
};

export type ThermoworksConnectionState =
  | "connecting"
  | "connected"
  | "disconnecting"
  | "disconnected";

export type ThermoworksConnectionReason =
  | "manual"
  | "unexpected"
  | "failed"
  | "shutdown"
  | "timeout"
  | "no_bluetooth"
  | "auth"
  | "unknown";

export type ThermoworksConnectionEvent = {
  type: "connection";
  state: ThermoworksConnectionState;
  deviceId: string | null;
  reason?: ThermoworksConnectionReason;
  message?: string;
};

export type ThermoworksReadingStatus = "ok" | "fault" | "unavailable";

export type ThermoworksReadingEvent = {
  type: "reading";
  deviceId: string;
  sensorId?: string;
  temperatureC: number | null;
  timestamp: number;
  sequence?: number;
  battery?: number;
  status: ThermoworksReadingStatus;
  raw?: {
    isReady?: boolean;
    isFault?: boolean;
    source?: string;
    notificationType?: string;
  };
};

export type ThermoworksErrorEvent = {
  type: "error";
  code: string;
  message: string;
  deviceId?: string;
};

export type ThermoworksButtonPressEvent = {
  type: "buttonPress";
  deviceId: string;
  timestamp: number;
};

export type ThermoworksScanResult = {
  ok: boolean;
  error?: string;
};

type ThermoworksEvents = {
  onDevices: (event: ThermoworksDevicesEvent) => void;
  onError: (event: ThermoworksErrorEvent) => void;
  onConnection: (event: ThermoworksConnectionEvent) => void;
  onReading: (event: ThermoworksReadingEvent) => void;
  onButtonPress: (event: ThermoworksButtonPressEvent) => void;
};

type AlenioThermoworksNativeModule = NativeModule<ThermoworksEvents> & {
  isAvailable(): boolean;
  initialize(): Promise<ThermoworksInitResult>;
  getDiagnostics(): Promise<ThermoworksDiagnostics>;
  ensureBluetoothPermissions(): Promise<ThermoworksScanResult>;
  startScan(): Promise<ThermoworksScanResult>;
  stopScan(): Promise<ThermoworksScanResult>;
  connect(deviceId: string): Promise<ThermoworksScanResult>;
  disconnect(): Promise<ThermoworksScanResult>;
  getDiscoveredDevices(): Promise<ThermoworksDiscoveredDevice[]>;
  addListener<EventName extends keyof ThermoworksEvents>(
    eventName: EventName,
    listener: ThermoworksEvents[EventName],
  ): EventSubscription;
};

let cached: AlenioThermoworksNativeModule | null | undefined;

/** Returns null when the native module is not present (e.g. Expo Go). */
export function getNativeOrNull(): AlenioThermoworksNativeModule | null {
  if (cached !== undefined) return cached;
  try {
    cached = requireNativeModule<AlenioThermoworksNativeModule>("AlenioThermoworks");
  } catch {
    cached = null;
  }
  return cached;
}

File: temps/modules/alenio-thermoworks/ios/AlenioThermoworksModule.swift
import ExpoModulesCore

public class AlenioThermoworksModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AlenioThermoworks")

    Events("onDevices", "onError", "onConnection")

    OnCreate {
      ThermaLibFacade.setDevicesBlock { devices in
        self.sendEvent("onDevices", [
          "type": "devices",
          "devices": devices
        ])
      }
      ThermaLibFacade.setErrorBlock { code, message in
        self.sendEvent("onError", [
          "type": "error",
          "code": code,
          "message": message
        ])
      }
      ThermaLibFacade.setConnectionBlock { event in
        self.sendEvent("onConnection", event)
      }
    }

    OnDestroy {
      ThermaLibFacade.shutdownDiscovery()
    }

    Function("isAvailable") { () -> Bool in
      return ThermaLibFacade.isLinked()
    }

    AsyncFunction("initialize") { () -> [String: Any?] in
      let result = ThermaLibFacade.initializeSdk()
      return [
        "ok": result["ok"] as? Bool ?? false,
        "sdkVersion": result["sdkVersion"] as? String ?? "",
        "bluetoothAvailable": result["bluetoothAvailable"] as? Bool ?? false,
        "error": result["error"] as? String
      ]
    }

    AsyncFunction("getDiagnostics") { () -> [String: Any?] in
      let version = ThermaLibFacade.sdkVersion()
      return [
        "module": "alenio-thermoworks",
        "platform": "ios",
        "sdkVersion": version,
        "available": ThermaLibFacade.isLinked(),
        "initialized": ThermaLibFacade.isInitialized(),
        "bluetoothAvailable": ThermaLibFacade.isBluetoothAvailable(),
        "scanning": ThermaLibFacade.isScanning(),
        "discoveredCount": ThermaLibFacade.discoveredDevices().count,
        "connectedDeviceId": ThermaLibFacade.connectedDeviceId(),
        "error": nil as String?
      ]
    }

    AsyncFunction("ensureBluetoothPermissions") { () -> [String: Any?] in
      let available = ThermaLibFacade.isBluetoothAvailable()
      return [
        "ok": available,
        "error": available ? nil : "Bluetooth is not available"
      ]
    }

    AsyncFunction("startScan") { () -> [String: Any?] in
      let result = ThermaLibFacade.startScan()
      return [
        "ok": result["ok"] as? Bool ?? false,
        "error": result["error"] as? String
      ]
    }

    AsyncFunction("stopScan") { () -> [String: Any?] in
      let result = ThermaLibFacade.stopScan()
      return [
        "ok": result["ok"] as? Bool ?? false,
        "error": result["error"] as? String
      ]
    }

    AsyncFunction("connect") { (deviceId: String) -> [String: Any?] in
      let result = ThermaLibFacade.connect(deviceId)
      return [
        "ok": result["ok"] as? Bool ?? false,
        "error": result["error"] as? String
      ]
    }

    AsyncFunction("disconnect") { () -> [String: Any?] in
      let result = ThermaLibFacade.disconnect()
      return [
        "ok": result["ok"] as? Bool ?? false,
        "error": result["error"] as? String
      ]
    }

    AsyncFunction("getDiscoveredDevices") { () -> [[String: Any]] in
      return ThermaLibFacade.discoveredDevices() as? [[String: Any]] ?? []
    }
  }
}
